### ISSUE TITLE
Exiting insert mode upon sending a message

### ISSUE THREAD

**Issue by 97-109-107** [GermainZ/weechat-vimode#24]
Is it a designed behaviour that the prompt doesn't exit INSERT mode upon writing a command/message and pressing ENTER? Is this an issue with my configuration/keys or perhaps can this behaviour can be toggled to be more vim-like?

_Thank you for the great plugin_

**Comment by GermainZ** [comment]
Add this below the `key_comma` function:

````
def key_return(buf, input_line, cur, count):
    """Return and go to insert mode."""
    weechat.command("", "/input return")
    set_mode("INSERT")
````

Just a bit below that, in ``VI_KEYS``, remove this line:

````
           '\x01M': "/input return",
````

… and replace it with:

````
           '\x01M': key_return,
````


I don't mind making it the default as I prefer this behavior. If someone has any complaints about this being the default in the next version, I can make an option for it.

**Comment by 97-109-107** [comment]
Thanks for the reply, I plugged it in as you described. That didn't work for me. The command ``    set_mode("INSERT")`` looked suspicious to me, so I changed it to NORMAL. That stil didn't work.

I reloaded weechat, and double checked that I have the proper file linked in there. 
Any clues?

**Comment by GermainZ** [comment]
Ah, I misread what you wanted. What I wrote above will automatically start insert mode after you send a message, if you were in normal mode. (For example, "Hey this is a test" Esc 4b D Enter would send "Hey" and go to insert mode instead of staying in normal mode.)

For entering normal mode automatically, the easiest way is something like this (you'll have to remap the enter key, which is ``^M``, in WeeChat; remember that you can use ``^J`` instead if anything goes wrong and vimode stops working):

Add this function after e.g. the `cb_vimode_cmd`` function:

````python
def cb_vimode_go_to_normal(data, buf, args):
    set_mode("NORMAL")
    return weechat.WEECHAT_RC_OK
````

… and add this at the very end of the script:

````python
    weechat.hook_command("vimode_go_to_normal", ("This command can be used for"
                         " key bindings to go to normal mode."), "", "", "",
                         "cb_vimode_go_to_normal", "")
````

… and finally remap ``^M`` to execute ``/vimode_go_to_normal`` followed by ``/input return`` or whatever Return is set to on your setup (to find out, press ``Alt-K`` followed by ``<Return>``):

````
/key bind ctrl-M /vimode_go_to_normal;/input return
````

**Comment by 97-109-107** [comment]
That's fantastic. It works the way I wanted now is definitely feels more intuitive (hah) for me at least.

I might try to wrap it into a commit with an option to toggle this behaviour, but how would you like to handle the remapping then? Would the behaviour of the enter key by now by default passing through the ``cb_vimode_go_to_normal`` (later triggering ``/input return``) which in turn checks for a configuration variable?

**Comment by GermainZ** [comment]
As a general rule, I don't touch user bindings automatically because I can't clean that up when the script is removed, so the user should always be aware of what happened in those cases.
It can be added to the FAQ but the user would have to remap ``^M`` manually.

**Comment by GermainZ** [comment]
Actually, we could also hook ``/input return`` and then have the check
there.

**Comment by 97-109-107** [comment]
I understand the logic, that's more along the lines what I was thinking above.

Would replacing this binding:
``'\x01M': "/input return"``
With a function that looks like:
````
def send_and_optionally_exit_insert()
  if return_to_normal
    set_mode("NORMAL")
  /input return
  return weechat.WEECHAT_RC_OK
````
Be the right direction?

**Comment by GermainZ** [comment]
VI_KEYS are only used when in normal mode, so that wouldn't work. I'll look
into this when I have some time.

**Comment by GermainZ** [comment]
Offering this as a config option is possible, but not everyone may be using `/input return` (e.g. multiline.pl users). That'd require an extra string config which will be have to be set by the user, which I think adds too much complexity for too little gain.

I've added an entry to the FAQ about this: [Exiting insert mode upon sending a message][1], and the necessary patch to allow has been merged (will be in the next version if you don't use weechat-vimode from GitHub). This is the best compromise I can think of. Cheers. :)

[1]: https://github.com/GermainZ/weechat-vimode/blob/master/FAQ.md#exiting-insert-mode-upon-sending-a-message
